gpflow/utilities/utilities.py

```diff
@@ -246,22 +247,123 @@ def clear_bijector(bijector, _, state):
     return input_module
 
 
-M = TypeVar('M', bound=tf.Module)
+def _get_by_name_index(parent: object, attr_str: str, index_str: str) -> object:
+    attr = getattr(parent, attr_str)
+    if index_str != "":
+        index = int(index_str)
```

Comment:
what if the index isn't an index by a dictionary key?

Reply:
Module path can have only lists or modules.

Reply:
looking at `traverse_module`, which is used indirectly via `parameter_dict`, it appears there is the possibility of dictionaries

Reply:
also nested lists are a concern. it's OK if we don't want to support them but we should be clear to users about that